### PR TITLE
Fix eigh gradients for large eigenvalues

### LINKED ISSUE #40141: `eigh` forward- and reverse-mode eigenvector gradients return all-`NaN` for eigenvalues above `2^p` (mantissa-precision cliff in the JVP rule's div-by-zero guard)
### Description

`jax.jacfwd`/`jax.jacrev` of `jnp.linalg.eigh`'s eigenvectors return an all-`NaN` Jacobian whenever the input matrix's eigenvalues exceed roughly `2^p`, where `p` is the number of significant mantissa bits for the dtype in use (`p=24` for float32, `p=53` for float64). PyTorch's `torch.linalg.eigh` produces the correct finite gradient for the identical input and dtype.

## Reproducible example

```python
import numpy as np, jax, jax.numpy as jnp, torch

base = np.float32(2**24)
lam = np.array([base, base + 2, base + 4], dtype=np.float32)
A = np.diag(lam)

f = lambda M: jnp.sum(jnp.linalg.eigh(M)[1])
g = lambda M: torch.sum(torch.linalg.eigh(M)[1])

print("JAX (float32):")
print(jax.jacfwd(f)(jnp.asarray(A)))

print("PyTorch (float32):")
print(torch.autograd.functional.jacobian(g, torch.tensor(A)))
```

## Output

```
JAX (float32):
[[nan nan nan]
 [nan nan nan]
 [nan nan nan]]

PyTorch (float32):
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
```

## Root cause

The eigenvector derivative for symmetric matrices with distinct eigenvalues is:

```
        ┌  1/(λ_j - λ_i)   i ≠ j
F_ij =  │
        └  0                i = j

dV = V (F ⊙ V^T dM V)
```

JAX's implementation avoids the literal `i=j` division-by-zero by adding the identity before inverting, then subtracting it back off, in [`_eigh_jvp_rule`, `jax/_src/lax/linalg.py`](https://github.com/jax-ml/jax/blob/main/jax/_src/lax/linalg.py#L1358-L1360).

```python
Fmat = lax.integer_pow(eye_n + w[..., np.newaxis, :] - w[..., np.newaxis], -1) - eye_n
```

Walking the diagonal (`i=j`) at eigenvalue magnitude `λ`:

1. `eye_n[i,i] + (λ - λ) = 1 + 0 = 1`, so that step 2 inverts a safe nonzero value.
2. But floating point does `1 + λ` as a single rounded operation: once `λ ≥ 2^p` (`ulp(λ) > 1`), the value `1` is smaller than the gap between adjacent representable floats near `λ`, so `1 + λ` rounds back down to exactly `λ`.
3. The subsequent `- λ` (from `w[..., None]`) then gives exactly `0`, not `1`.
4. `1/0 = inf`, then `inf - 1 = inf`.
5. This `inf` is multiplied elementwise against the correctly-computed `0` entry of `V^T dM V` on the diagonal, and the resulting `NaN` propagates through the final `dot(v, Fmat * vdag_adot_v)` into every entry of the output.

Because reverse-mode differentiation of `eigh` is obtained by transposing this same JVP rule registered, `jax.jacrev` fails identically to `jax.jacfwd`.

## Why PyTorch is unaffected

PyTorch never derives the diagonal from an arithmetic identity. It assigns the
diagonal outright.

In `linalg_eig_backward`([`FunctionsManual.cpp#L4130-L4151`](https://github.com/pytorch/pytorch/blob/f5f7ac4f0d1deffae10abcb959808c80944119a5/torch/csrc/autograd/FunctionsManual.cpp#L4130-L4151)):

```cpp
auto Econj = [&L] {
  auto Lconj = L.conj();
  auto ret = Lconj.unsqueeze(-2) - Lconj.unsqueeze(-1);
  ret.diagonal(0, -2, -1).fill_(1.);       
  return ret;
}();

auto ret = VhgV.div_(Econj);              

if (gL.defined()) {
  // ...
  ret.diagonal(0, -2, -1).copy_(gL);       
}
```

First, `fill_(1.)` is a plain store into the diagonal
rather than the sum `λ + 1 - λ`. Second, the
placeholder is irrelevant regardless: L4147 throws away the entire diagonal of the quotient and replaces it with `gL`. Similarily for forward mode.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.6.2
jaxlib: 0.6.2
numpy:  2.2.6
python: 3.10.19 (main, Oct 21 2025, 16:37:10) [Clang 20.1.8 ]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', node='xx', release='25.6.0', version='Darwin Kernel Version 25.6.0: Fri Jul 31 19:18:29 PDT 2026; root:xnu-12377.161.14~5/RELEASE_ARM64_T8122', machine='arm64')
```

### PR DESCRIPTION
**Pull request by edluyuan**
Fixes #40141.

`_eigh_jvp_rule` constructs the reciprocal eigenvalue-gap matrix using
```
    eye + w_j - w_i
```
to avoid dividing by zero on its diagonal.

Although this is correct in exact arithmetic, the evaluation order is
numerically unsafe for large eigenvalues. 

For example, at `2**24` in
float32, adding `1` to an eigenvalue may round back to the same
floating-point value. The subsequent subtraction can therefore produce
zero (or another incorrect value) on the diagonal, causing the
reciprocal to produce non-finite or incorrect entries.

Compute the eigenvalue differences first and add the identity guard
afterward:
```
    (w_j - w_i) + eye
```
Diagonal subtraction is then exactly zero before the guard is
added, while off-diagonal eigenvalue differences are unchanged.

Add a regression test using distinct float32 eigenvalues around
`2**24`, checking that both `jacfwd` and `jacrev` of the eigenvectors
remain finite.

Tests passed:
- `pytest tests/eigh_test.py -k EighGradLargeEigenvalues`
- `pytest tests/eigh_test.py -k EighGrad`
- `pytest tests/eigh_test.py`
- `JAX_ENABLE_X64=1 pytest tests/eigh_test.py -k EighGrad`

### CODE REVIEW

**Review by CAOShurong — COMMENTED** [review]
## Description

Verified end-to-end on Windows (CPython 3.13.1, CPU jaxlib) against the exact PR head `cd87a6bc854c1cf60a4d2d552f5102a1cba21d64` and current main (`8c191b842df0`). The diagnosis is correct, and the fix is the minimal right one:

**Root cause confirmed.** In `_eigh_jvp_rule`, the old expression `eye + w[..., newaxis, :] - w[..., :, newaxis]` evaluates `eye + w_j` *before* subtracting `w_i`. At eigenvalue magnitudes ≥ 2²⁴ float32 can't represent `w_j + 1`, so the diagonal term collapses to a nonzero value that survives the subtraction — I reproduced this directly by evaluating both expressions at `w = [2**24, 2**24+2, 2**24+4]`:

- old Fmat diag: `[inf, -0.5, inf]` (two of three diagonal entries wrong)
- new Fmat diag: `[0., 0., 0.]` (exact zeros before the guard, as intended)
- off-diagonal entries are **bit-identical** between old and new (max abs diff 0.0), confirming the reorder only touches the guard path.

**Fix verified at PR head.** `testEighGradLargeEigenvalues` passes for both `jacfwd` and `jacrev`; full `tests/eigh_test.py`: **74 passed, 21 skipped**. `ruff check` clean on both changed files.

**Regression test is well-constructed**: distinct eigenvalues spaced exactly at the float32 ULP boundary around 2²⁴, checking finiteness through both jacobian paths.

One small note for completeness: with complex Hermitian input, `delta_w + eye_n` still relies on the imaginary parts cancelling exactly on the diagonal — which they do here since `delta_w`'s diagonal is computed as an exact `w_i - w_i`. So the fix holds for the complex path too.

LGTM.

**Comment by edluyuan** [comment]
Could a maintainer add the review label when convenient? Thanks！

**Review by jakevdp — APPROVED** [review]
Looks good, thanks!

**Comment by edluyuan** [comment]
Hi @jakevdp It looks like the internal Copybara checks failed after the import. Would you mind rerunning the internal checks when you get a chance? Thanks!
